supervisor/supervisor.go

```diff
@@ -0,0 +1,311 @@
+// to lines of output from the job
+type logWatchers struct {
+	*sync.RWMutex
+	watchers map[chan LogOutput]bool
+}
+
+// newLogWatcher returns an instance of a logWatchers
+func newLogWatcher() *logWatchers {
+	return &logWatchers{
+		RWMutex:  &sync.RWMutex{},
+		watchers: make(map[chan LogOutput]bool),
+	}
+}
+
+// addWatcher adds the provided channel to the logWatchers set
+func (l *logWatchers) addWatcher(c chan LogOutput) {
+	l.Lock()
+	defer l.Unlock()
+
+	l.watchers[c] = true
+}
+
+// sendLog sends the log message to all the subscribers in the logWatchers set
+func (l *logWatchers) sendLog(log LogOutput) {
+	l.RLock()
+	defer l.RUnlock()
+
+	for c := range l.watchers {
+		c <- log
+	}
+}
+
+// removeLogWatcher removes the channel from the logWatchers set
+func (l *logWatchers) removeLogWatcher(c chan LogOutput) {
+	l.Lock()
+	defer l.Unlock()
+
+	delete(l.watchers, c)
+}
+
+// Start starts a job and returns the current status. An ErrExecNotFound
+// is returned if the executable cannot be found by the system.
+// A unique id is generated for the job and returned in JobStatus
+func (s *Supervisor) Start(cmd string, args []string) (*JobStatus, error) {
+	if cmd == "" {
+		msg := "cannot add job, no executable defined"
+		s.logger.Error(msg)
+		return nil, errors.New(msg)
+	}
+
+	_, err := exec.LookPath(cmd)
+
+	if err != nil {
+		msg := fmt.Sprintf("cannot add job, the system cannot find the given executable \"%s\"", cmd)
+		s.logger.Error(msg)
+		return nil, ErrExecNotFound
+	}
+
+	jobId := uuid.New().String()
+	p := NewProcess(cmd, args)
+	sc := p.Start()
+
+	s.wg.Add(1)
+	go func(id string, p *Process, c <-chan Status) {
+		defer s.wg.Done()
+		s.run(id, p, sc)
+	}(jobId, p, sc)
+
+	s.logger.Info("Added Job: ",
+		zap.String("id", jobId),
+		zap.String("name", cmd),
+		zap.Strings("args", args))
+
+	s.access.Lock()
+	s.jobs.Store(jobId, p)
+	s.outputChannels[jobId] = newLogWatcher()
+	s.ouputBuffers[jobId] = []LogOutput{}
+	s.access.Unlock()
+
+	status := p.GetStatus()
+
+	return &JobStatus{jobId, status}, nil
+}
+
+// run listens for output from the running job through it's StdOut and StdErr channels.
+// Each line of output is first added to the job's line buffer before it's sent to the log watchers.
+func (s *Supervisor) run(id string, proc *Process, sc <-chan Status) {
+	go func(id string, p *Process) {
+		for {
+			select {
+			case l := <-p.StdOut():
+				line := LogOutput{StdOut, l}
+
+				s.access.Lock()
+				s.ouputBuffers[id] = append(s.ouputBuffers[id], line)
+				watchers := s.outputChannels[id]
+				s.access.Unlock()
+
+				watchers.sendLog(line)
+			case l := <-p.StdErr():
+				line := LogOutput{StdErr, l}
+
+				s.access.Lock()
+				s.ouputBuffers[id] = append(s.ouputBuffers[id], line)
+				watchers := s.outputChannels[id]
+				s.access.Unlock()
+
+				watchers.sendLog(line)
+			case <-p.Done():
+				return
+			}
+
+		}
+	}(id, proc)
+
+	fs := <-sc
```

Comment:
Is this the main use of the status chan? Could you just receive from `p.Done()`, lock p.status, and then read the status?